synapse/http/client.py

```diff
@@ -81,13 +115,36 @@ def __init__(self, hs):
         )
         self.user_agent = hs.version_string
         self.clock = hs.get_clock()
+        self.reactor = hs.get_reactor()
         if hs.config.user_agent_suffix:
-            self.user_agent = "%s %s" % (self.user_agent, hs.config.user_agent_suffix,)
+            self.user_agent = "%s %s" % (self.user_agent, hs.config.user_agent_suffix)
 
         self.user_agent = self.user_agent.encode('ascii')
 
     @defer.inlineCallbacks
     def request(self, method, uri, data=b'', headers=None):
+
+        # Check our IP whitelists/blacklists before making the request.
+        if self.blacklist:
+            host = URI.fromBytes(uri.encode('utf8')).host.decode('utf8')
+
+            # Fetch all the possible IP addresses.
+            resolver_target = self.reactor.nameResolver.resolveHostName(
+                IPListResolutionReceiver(), host
+            )
+            addresses = yield resolver_target.get_deferred()
+
+            for i in addresses:
+                ip_address = IPAddress(i.host)
+
+                if ip_address in self.blacklist:
+                    if self.whitelist is None or ip_address not in self.whitelist:
+                        raise SynapseError(
```

Comment:
I think it would be helpful to log something here, including the hostname being resolved and the IP address being blocked.